File: Emerging_programming_paradigms/Codice/Python/adt_expr.py
## Si vuole rappresentare con una struttura dati gli alberi di sintassi astratta 
## delle espressioni di un linguaggio di programmazione.

from __future__ import annotations
from dataclasses import dataclass

@dataclass
class Num:
    n: int

@dataclass
class String:
    n: str

@dataclass
class Mult:
    e1: Expr
    e2: Expr

@dataclass
class Concat:
    e1: Expr
    e2: Expr

@dataclass
class Eq:
    e1: Expr
    e2: Expr

Expr = Num | String | Mult | Concat | Eq

def eval(e):
    match e:
        case Num(n):
            return n
        case String(s):
            return s
        case Mult(e1, e2):
            return eval(e1) * eval(e2)
        case Concat(e1, e2):
            return eval(e1) + eval(e2)
        case Eq(e1, e2):
            return eval(e1) == eval(e2)

good = Eq(Mult(Num(3), Num(4)), Num(5))
bad = Eq(Concat(String("ciao"), Num(2)), Num(3))

print(f"Good: {eval(good)}")
print(f"Bad: {eval(bad)}")

File: Emerging_programming_paradigms/Codice/OCaml/adt_expr.ml
type expr =
    Num : int -> expr
  | Bool : bool -> expr
  | Mult : expr * expr -> expr
  | And : expr * expr -> expr
  | Eq : expr * expr -> expr

(* Es: Mult (Num 3) (Bool true) è rappresentata come { Mult, { Num, 3 }, { Bool true } } *)

type res = 
    | I : int -> res
    | B : bool -> res
    | E : res

let to_string =
    function
        I n -> string_of_int n
    |   B b -> string_of_bool b
    |   E -> "error"

let rec eval: expr -> ??? =
    function
        Num n -> I n
    |   Bool b -> B b
    |   Mult(e1, e2) ->
            (match eval e1, eval e2 with
             | I n1, I n2 -> I (n1 * n2)
             | _, _ -> E)
    |   And(e1, e2) ->
            (match eval e1, eval e2 with
             | B b1, B b2 -> B (b1 && b2)
             | _, _ -> E)
    |   Eq(e1, e2) ->
            (match eval e1, eval e2 with
             | I n1, I n2 -> B (n1 == n2)
             | B b1, B b2 -> B (b1 == b2)
             | _, _ -> E)

let good = Eq (Mult (Num 3, Num 4), Num 5)
let bad = Eq (And (Bool true, Num 4), Num 5)

let _ = 
    Printf.printf "%s\n" (to_string (eval good)) ;
    Printf.printf "%s\n" (to_string (eval bad))